### ISSUE TITLE
done{T,S}(r::StepRange{T,S}, i::Integer) is problematic

### ISSUE THREAD

**Issue by ajkeller34** [JuliaLang/julia#14257]
In `base/range.jl`, the following method is problematic:

```done{T,S}(r::StepRange{T,S}, i::Integer) =
    isempty(r) | (i == oftype(i, r.stop) + r.step)```

Consider the following for loop, which exits immediately:

```
for x in 0x0000:0x0010:0xFFF0
    println(x)  # never gets called
end
```

The problem is due to an overflow of UInt16, but without knowing the iterator implementation the end user has no idea an overflow would occur. The more general method does not have this problem:

```done{T,S}(r::StepRange{T,S}, i) = isempty(r) | (i < min(r.start, r.stop)) | (i > max(r.start, r.stop))```

**Comment by JeffBezanson** [comment]
Fixed by 7b22efefa6d3e28e8c6063ac2806d4233822acd7 I think?

**Comment by ajkeller34** [comment]
Thanks for the link, I see that it is correct in master. I'll check that in the future before opening issues (though a fix would be appreciated in an any upcoming 0.4 release).
